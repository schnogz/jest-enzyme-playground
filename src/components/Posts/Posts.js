import React from 'react';
import styled from 'styled-components';
import HoverLink from './../HoverLink/HoverLink';
import Api from './../../services/api';
import Post from './../Post/Post';

const Wrapper = styled.section`
  margin: .4em;
  padding: 2em;
  background: papayawhip;
`;

const List = styled.div`
	width:100%;
	height:100%;
`;

class Posts extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			loaded: false,
			posts: []
		};
	}

	componentDidMount() {
		Api.getPosts()
			.then(data => {
				const posts = data.data.children.map(obj => obj.data);
				this.setState({posts: posts, loaded: true});
			});
	}

	render() {
		return (
			<Wrapper>
				<HoverLink page='https://www.reddit.com/r/bitcoin/new.json?sort=top'>Hot /r/bitcoin Posts</HoverLink>
				<List>
					{this.state.posts.map(post =>
						<Post key={post.id} post={post}/>
					)}
				</List>
			</Wrapper>
		);
	}
}

export default Posts;