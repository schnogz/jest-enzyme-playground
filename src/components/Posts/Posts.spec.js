import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json';

import Posts from './Posts';

describe('Posts component', () => {
	describe('using __mocks__', () => {
		// use __mocks__ to fake network calls
		jest.mock('./../../services/api');

		it('tree renders correctly', () => {
			const component = shallow(<Posts/>);
			const tree = toJson(component);
			expect(tree).toMatchSnapshot();
		});

		it('calls componentDidMount() lifecycle method', () => {
			jest.spyOn(Posts.prototype, 'componentDidMount');

			const component = mount(<Posts />);

			expect(Posts.prototype.componentDidMount).toHaveBeenCalled();
		});
	});

	describe('using inline mock', () => {

		jest.mock('./../../services/api', () => {
			return {
				getPosts: () => { return new Promise((resolve, reject) => {
					process.nextTick(
						() => resolve({ data: {
							children: [
								{ name: 'child 1' }
							]
						}}),
						() => reject({ error: 'something went wrong' })
					);
				});
				}
			}
		});

		it('tree renders correctly', () => {
			const component = shallow(<Posts/>);
			const tree = toJson(component);
			expect(tree).toMatchSnapshot();
		});

		it('calls componentDidMount() lifecycle method', () => {
			jest.spyOn(Posts.prototype, 'componentDidMount');

			const component = mount(<Posts />);

			expect(Posts.prototype.componentDidMount).toHaveBeenCalled();
		});
	});
});