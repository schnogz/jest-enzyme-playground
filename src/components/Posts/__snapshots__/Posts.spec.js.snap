// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component using __mocks__ tree renders correctly 1`] = `
<styled.section>
  <HoverLink
    page="https://www.reddit.com/r/bitcoin/new.json?sort=top"
  >
    Hot /r/bitcoin Posts
  </HoverLink>
  <styled.div />
</styled.section>
`;

exports[`Posts component using inline mock tree renders correctly 1`] = `
<styled.section>
  <HoverLink
    page="https://www.reddit.com/r/bitcoin/new.json?sort=top"
  >
    Hot /r/bitcoin Posts
  </HoverLink>
  <styled.div />
</styled.section>
`;
